src/traces/streamtube/calc.js

```diff
@@ -60,44 +61,95 @@ module.exports = function calc(gd, trace) {
     var filledX;
     var filledY;
     var filledZ;
-    var prevX;
-    var prevY;
-    var prevZ;
+    var firstX;
+    var firstY;
+    var firstZ;
     if(len) {
-        prevX = x[0];
-        prevY = y[0];
-        prevZ = z[0];
+        firstX = x[0];
+        firstY = y[0];
+        firstZ = z[0];
     }
 
     for(i = 0; i < len; i++) {
-        var xx = x[i];
-        xMax = Math.max(xMax, xx);
-        xMin = Math.min(xMin, xx);
+        xMax = Math.max(xMax, x[i]);
+        xMin = Math.min(xMin, x[i]);
 
-        var yy = y[i];
-        yMax = Math.max(yMax, yy);
-        yMin = Math.min(yMin, yy);
+        yMax = Math.max(yMax, y[i]);
+        yMin = Math.min(yMin, y[i]);
 
-        var zz = z[i];
-        zMax = Math.max(zMax, zz);
-        zMin = Math.min(zMin, zz);
+        zMax = Math.max(zMax, z[i]);
+        zMin = Math.min(zMin, z[i]);
 
-        if(!filledX && xx !== prevX) {
+        if(!filledX && x[i] !== firstX) {
             filledX = true;
             gridFill += 'x';
-        } else if(!filledY && yy !== prevY) {
+        } else if(!filledY && y[i] !== firstY) {
             filledY = true;
             gridFill += 'y';
-        } else if(!filledZ && zz !== prevZ) {
+        } else if(!filledZ && z[i] !== firstZ) {
             filledZ = true;
             gridFill += 'z';
         }
     }
-    // fill if not filled - case of having 1 dimension
+    // fill if not filled - case of having dimension(s) with one item
     if(!filledX) gridFill += 'x';
     if(!filledY) gridFill += 'y';
     if(!filledZ) gridFill += 'z';
 
+    var valsx = distinctVals(trace.x.slice(0, len));
+    var valsy = distinctVals(trace.y.slice(0, len));
+    var valsz = distinctVals(trace.z.slice(0, len));
+
+    var getArray = function(c) { return c === 'x' ? x : c === 'y' ? y : z; };
+    var getVals = function(c) { return c === 'x' ? valsx : c === 'y' ? valsy : valsz; };
+    var getLength = function(c) { return getVals(c).length; };
+    var getDir = function(c) { return (+(c[c.length - 1] - c[0])) * 2 + 1; };
```

Comment:
Hmm. Can you rewrite this. Turning `xyz` into `+1` or `-1` is a little weird.

Reply:
Good call. Done in https://github.com/plotly/plotly.js/pull/4271/commits/4af498408f1dcd8fdbb746667243160c2041fa56.